doc/update.md

```diff
@@ -0,0 +1,94 @@
+  bg_color?: number // 32-bit RGBA value, default 0
+  weight?: number // 100..900, default 400
+  italic?: boolean  // default false
+```
+
+It's not hard to imagine more style properties (typeface, size, OpenType features, etc).
+
+The guarantee on id is that it is not currently in use in any lines in the view. However, in practice, it will probably just count up. It can also be assumed to be small, so using it as an index into a dense array is reasonable.
+
+Discussion question: should the scope of set_style be to a tab, or to the global session?
+
+Style number 0 is reserved for selections. Discussion question: should other styles be reserved, like 1 for find results?
+
+```
+scroll_to: [number, number]  // line, column (in utf-8 code units)
+```
+
+```
+update
+  rev: number
+  ops: Op[]
+
+interface Op {
+  op: "copy" | "skip" | "invalidate" | "update" | "ins"
+  n: number  // number of lines affected
+  lines?: Line[]  // only present when op is "update" or "ins"  
+}
+```
+
+An update request can be seen as a function from the old client cache state to a new one. During evaluation, maintain an index (`old_ix`) into the old `lines` array, initially 0, and a new lines array, initially empty. [Note that this document specifies the semantics. The actual implementation will almost certainly represent at least initial and trailing sequences of invalid lines by their count; and the editing operations may be more efficiently done in-place than by copying from the old state to the new].
+
+The "copy" op appends the `n` lines `[old_ix: old_ix + n]` to the new lines array, and increments `old_ix` by `n`.
+
+The "skip" op increments `old_ix` by `n`.
+
+The "invalidate" op appends n invalid lines to the new lines array.
+
+The "ins" op appends new lines, specified by the "`lines`" parameter, specified in more detail below. For this op, `n` must equal `lines.length` (alternative: make n optional in this case). It does not update `old_ix`.
+
+The "update" op updates the cursor and/or style of n existing lines. As in "ins", n must equal lines.length. It also increments `old_ix` by `n`.
+
+In all cases, n is guaranteed positive and nonzero (as a consequence, any line present in the old state is copied at most once to the new state).
+
+```
+interface Line {
+  text?: string  // present when op is "update"
+  cursor?: number[]  // utf-8 code point offsets, in increasing order
+  styles?: number[]  // length is a multiple of 3, see below
+}
+```
+
+The interpretation of a line is different for "update" or "ins" ops. In an "ins" op, text is always present, and missing cursor or styles properties are interpreted as empty (no cursors on that line, no styles).
+
+In an "update" op, then the text property is absent from the line, and text is copied from the previous state (or left invalid if the previous state is invalid), and the cursor and styles are updated if present. To delete cursors from a line, the core sets the cursor property to the empty list.
```

Comment:
I am not sure about the meaning of the first sentence. Is it maybe supposed to be:

> In an "update" op, **when** the text property is absent from the line, **then** text is copied from the previous state (or left invalid if the previous state is invalid), and the cursor and styles are updated if present.